server/public/model/packet_metadata.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	"gopkg.in/yaml.v2"
+)
+
+type PacketType string
+
+const (
+	CurrentMetadataVersion int        = 1
+	SupportPacketType      PacketType = "support-packet"
+	PluginPacketType       PacketType = "plugin-packet"
+)
+
+type PacketMetadata struct {
+	// Required Fields
+
+	Version       int        `yaml:"version"`
+	Type          PacketType `yaml:"type"`
+	GeneratedAt   int64      `yaml:"generated_at"`
+	ServerVersion string     `yaml:"server_version"`
+	ServerID      string     `yaml:"server_id"`
+
+	// Optional Fields
+
+	LicenseID  string         `yaml:"license_id"`
+	CustomerID string         `yaml:"customer_id"`
+	Extras     map[string]any `yaml:"extras,omitempty"`
+}
+
+func (md *PacketMetadata) Validate() error {
+	if md.Version < 1 {
+		return fmt.Errorf("metadata version should be greater than 1")
+	}
+
+	switch md.Type {
+	case SupportPacketType, PluginPacketType:
+	default:
+		return fmt.Errorf("unrecognized packet type: %s", md.Type)
+	}
+
+	if _, err := semver.ParseTolerant(md.ServerVersion); err != nil {
+		return fmt.Errorf("could not parse server version: %w", err)
+	}
+
+	if !IsValidId(md.ServerID) {
+		return fmt.Errorf("server id is not a valid id %q", md.ServerID)
+	}
+
+	if md.LicenseID != "" && !IsValidId(md.LicenseID) {
+		return fmt.Errorf("license id is not a valid id %q", md.LicenseID)
+	}
+
+	if md.CustomerID != "" && !IsValidId(md.CustomerID) {
+		return fmt.Errorf("customer id is not a valid id %q", md.CustomerID)
+	}
+
+	return nil
+}
+
+func ParsePacketMetadata(b []byte) (*PacketMetadata, error) {
+	v := struct {
+		Version int `yaml:"version"`
+	}{}
+
+	err := yaml.Unmarshal(b, &v)
+	if err != nil {
+		return nil, err
+	}
+
+	switch v.Version {
+	case 1:
+		var md PacketMetadata
+		err = yaml.Unmarshal(b, &md)
+		if err != nil {
+			return nil, err
+		}
+
+		err = md.Validate()
+		if err != nil {
+			return nil, err
+		}
+
+		return &md, nil
+	default:
+		return nil, fmt.Errorf("unsupported metadata version: %d", v.Version)
+	}
+}
+
+// GeneratePacketMetadata is a utility function to generate metadata for customer provided Packets.
+// It will construct it from serverID and optionally a license.
+func GeneratePacketMetadata(serverID string, license *License, extra map[string]any) (*PacketMetadata, error) {
+	if extra == nil {
+		extra = make(map[string]any)
+	}
+
+	md := PacketMetadata{
+		Version:       CurrentMetadataVersion,
+		Type:          PluginPacketType,
+		GeneratedAt:   GetMillis(),
+		ServerVersion: CurrentVersion,
+		ServerID:      serverID,
+		Extras:        extra,
+	}
+
+	if license != nil {
```

Comment:
Not really sure about thus, but if license is missing then `LicenseId` and `CustomerID` will be null, not empty string. Either of them should be OK, as long as the decision is finalized.

Reply:
In Go, the zero value (i.e. the initial value) of `string`s is the empty string. They can only be `null` if the type is a string pointer, i.e. `*string`, but they can't if a plain string is used.